posthog/ai/langchain/callbacks.py

```diff
@@ -31,26 +32,38 @@
 log = logging.getLogger("posthog")
 
 
-class RunMetadata(TypedDict, total=False):
-    input: Any
-    """Input of the run: messages, prompt variables, etc."""
+@dataclass
+class SpanMetadata:
     name: str
     """Name of the run: chain name, model name, etc."""
-    provider: str
+    start_time: float
+    """Start time of the run."""
+    end_time: Optional[float]
+    """End time of the run."""
+    input: Optional[Any]
+    """Input of the run: messages, prompt variables, etc."""
+
+    @property
+    def latency(self) -> float:
+        if not self.end_time:
+            return 0
+        return self.end_time - self.start_time
+
+
+@dataclass
+class GenerationMetadata(SpanMetadata):
+    provider: Optional[str] = None
     """Provider of the run: OpenAI, Anthropic"""
-    model: str
+    model: Optional[str] = None
     """Model used in the run"""
-    model_params: Dict[str, Any]
+    model_params: Optional[Dict[str, Any]] = None
     """Model parameters of the run: temperature, max_tokens, etc."""
-    base_url: str
+    base_url: Optional[str] = None
     """Base URL of the provider's API used in the run."""
-    start_time: float
-    """Start time of the run."""
-    end_time: float
-    """End time of the run."""
 
 
-RunStorage = Dict[UUID, RunMetadata]
+RunMetadataUnion = Union[SpanMetadata, GenerationMetadata]
```

Comment:
Nit: A little redundant to suffix with `Union`
```suggestion
RunMetadata = Union[SpanMetadata, GenerationMetadata]
```

Reply:
Oh yeah, I've missed that. There was a naming conflict before haha. I'll fix it!